### ISSUE TITLE
Memory leak when rebalancing repetition trees

### ISSUE THREAD

**Issue by philipturnbull** [tree-sitter/tree-sitter#131]
Fuzzing with AddressSanitizer uncovered a memory leak which can be triggered by the bash, C and python grammars:
```
=================================================================
==15==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 128 byte(s) in 1 object(s) allocated from:
    #0 0x4dd168 in malloc /src/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:88
    #1 0x568d9b in ts_malloc /src/octofuzz/src/runtime/alloc.h:49:18
    #2 0x56aa4b in ts_tree_pool_allocate /src/octofuzz/src/runtime/tree.c:151:12
    #3 0x56abf8 in ts_tree_make_leaf /src/octofuzz/src/runtime/tree.c:167:18
    #4 0x56dd04 in ts_tree_make_node /src/octofuzz/src/runtime/tree.c:378:18
    #5 0x55fbd0 in parser__reduce /src/octofuzz/src/runtime/parser.c:665:20
    #6 0x55bf87 in parser__advance /src/octofuzz/src/runtime/parser.c:1123:38
    #7 0x55adca in parser_parse /src/octofuzz/src/runtime/parser.c:1234:9
    #8 0x55352f in ts_document_parse_with_options /src/octofuzz/src/runtime/document.c:137:16
    #9 0x5182eb in LLVMFuzzerTestOneInput /src/octofuzz/test/fuzz/fuzzer.cc:21:3
    #10 0x5a470e in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/libfuzzer/FuzzerLoop.cpp:463:13
    #11 0x584145 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) /src/libfuzzer/FuzzerDriver.cpp:273:6
    #12 0x58f3df in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/libfuzzer/FuzzerDriver.cpp:689:9
    #13 0x5837e8 in main /src/libfuzzer/FuzzerMain.cpp:20:10
    #14 0x7f184626682f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
```

<details>

```
Indirect leak of 512 byte(s) in 4 object(s) allocated from:
    #0 0x4dd168 in malloc /src/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:88
    #1 0x568d9b in ts_malloc /src/octofuzz/src/runtime/alloc.h:49:18
    #2 0x56aa4b in ts_tree_pool_allocate /src/octofuzz/src/runtime/tree.c:151:12
    #3 0x56abf8 in ts_tree_make_leaf /src/octofuzz/src/runtime/tree.c:167:18
    #4 0x564823 in parser__lex /src/octofuzz/src/runtime/parser.c:427:14
    #5 0x55e701 in parser__get_lookahead /src/octofuzz/src/runtime/parser.c:556:12
    #6 0x55bc07 in parser__advance /src/octofuzz/src/runtime/parser.c:1090:21
    #7 0x55adca in parser_parse /src/octofuzz/src/runtime/parser.c:1234:9
    #8 0x55352f in ts_document_parse_with_options /src/octofuzz/src/runtime/document.c:137:16
    #9 0x5182eb in LLVMFuzzerTestOneInput /src/octofuzz/test/fuzz/fuzzer.cc:21:3
    #10 0x5a470e in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/libfuzzer/FuzzerLoop.cpp:463:13
    #11 0x584145 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) /src/libfuzzer/FuzzerDriver.cpp:273:6
    #12 0x58f3df in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/libfuzzer/FuzzerDriver.cpp:689:9
    #13 0x5837e8 in main /src/libfuzzer/FuzzerMain.cpp:20:10
    #14 0x7f184626682f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

Indirect leak of 384 byte(s) in 3 object(s) allocated from:
    #0 0x4dd168 in malloc /src/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:88
    #1 0x568d9b in ts_malloc /src/octofuzz/src/runtime/alloc.h:49:18
    #2 0x56aa4b in ts_tree_pool_allocate /src/octofuzz/src/runtime/tree.c:151:12
    #3 0x56abf8 in ts_tree_make_leaf /src/octofuzz/src/runtime/tree.c:167:18
    #4 0x56dd04 in ts_tree_make_node /src/octofuzz/src/runtime/tree.c:378:18
    #5 0x55fbd0 in parser__reduce /src/octofuzz/src/runtime/parser.c:665:20
    #6 0x567fa1 in parser__do_all_potential_reductions /src/octofuzz/src/runtime/parser.c:837:7
    #7 0x562a6f in parser__handle_error /src/octofuzz/src/runtime/parser.c:915:15
    #8 0x55c2c0 in parser__advance /src/octofuzz/src/runtime/parser.c:1162:7
    #9 0x55adca in parser_parse /src/octofuzz/src/runtime/parser.c:1234:9
    #10 0x55352f in ts_document_parse_with_options /src/octofuzz/src/runtime/document.c:137:16
    #11 0x5182eb in LLVMFuzzerTestOneInput /src/octofuzz/test/fuzz/fuzzer.cc:21:3
    #12 0x5a470e in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/libfuzzer/FuzzerLoop.cpp:463:13
    #13 0x584145 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) /src/libfuzzer/FuzzerDriver.cpp:273:6
    #14 0x58f3df in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/libfuzzer/FuzzerDriver.cpp:689:9
    #15 0x5837e8 in main /src/libfuzzer/FuzzerMain.cpp:20:10
    #16 0x7f184626682f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

Indirect leak of 192 byte(s) in 3 object(s) allocated from:
    #0 0x4dd380 in calloc /src/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:97
    #1 0x5737d4 in ts_calloc /src/octofuzz/src/runtime/alloc.h:58:18
    #2 0x573934 in array__grow /src/octofuzz/src/runtime/array.h:93:24
    #3 0x57732b in stack__iter /src/octofuzz/src/runtime/stack.c:299:13
    #4 0x57732b in ts_stack_pop_count /src/octofuzz/src/runtime/stack.c:427
    #5 0x55f96c in parser__reduce /src/octofuzz/src/runtime/parser.c:652:24
    #6 0x567fa1 in parser__do_all_potential_reductions /src/octofuzz/src/runtime/parser.c:837:7
    #7 0x562a6f in parser__handle_error /src/octofuzz/src/runtime/parser.c:915:15
    #8 0x55c2c0 in parser__advance /src/octofuzz/src/runtime/parser.c:1162:7
    #9 0x55adca in parser_parse /src/octofuzz/src/runtime/parser.c:1234:9
    #10 0x55352f in ts_document_parse_with_options /src/octofuzz/src/runtime/document.c:137:16
    #11 0x5182eb in LLVMFuzzerTestOneInput /src/octofuzz/test/fuzz/fuzzer.cc:21:3
    #12 0x5a470e in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/libfuzzer/FuzzerLoop.cpp:463:13
    #13 0x584145 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) /src/libfuzzer/FuzzerDriver.cpp:273:6
    #14 0x58f3df in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/libfuzzer/FuzzerDriver.cpp:689:9
    #15 0x5837e8 in main /src/libfuzzer/FuzzerMain.cpp:20:10
    #16 0x7f184626682f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

Indirect leak of 128 byte(s) in 1 object(s) allocated from:
    #0 0x4dd168 in malloc /src/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:88
    #1 0x568d9b in ts_malloc /src/octofuzz/src/runtime/alloc.h:49:18
    #2 0x56aa4b in ts_tree_pool_allocate /src/octofuzz/src/runtime/tree.c:151:12
    #3 0x56abf8 in ts_tree_make_leaf /src/octofuzz/src/runtime/tree.c:167:18
    #4 0x56e335 in ts_tree_make_missing_leaf /src/octofuzz/src/runtime/tree.c:408:18
    #5 0x562a1f in parser__handle_error /src/octofuzz/src/runtime/parser.c:907:32
    #6 0x55c2c0 in parser__advance /src/octofuzz/src/runtime/parser.c:1162:7
    #7 0x55adca in parser_parse /src/octofuzz/src/runtime/parser.c:1234:9
    #8 0x55352f in ts_document_parse_with_options /src/octofuzz/src/runtime/document.c:137:16
    #9 0x5182eb in LLVMFuzzerTestOneInput /src/octofuzz/test/fuzz/fuzzer.cc:21:3
    #10 0x5a470e in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/libfuzzer/FuzzerLoop.cpp:463:13
    #11 0x584145 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) /src/libfuzzer/FuzzerDriver.cpp:273:6
    #12 0x58f3df in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/libfuzzer/FuzzerDriver.cpp:689:9
    #13 0x5837e8 in main /src/libfuzzer/FuzzerMain.cpp:20:10
    #14 0x7f184626682f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

Indirect leak of 128 byte(s) in 2 object(s) allocated from:
    #0 0x4dd380 in calloc /src/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:97
    #1 0x5737d4 in ts_calloc /src/octofuzz/src/runtime/alloc.h:58:18
    #2 0x573934 in array__grow /src/octofuzz/src/runtime/array.h:93:24
    #3 0x57732b in stack__iter /src/octofuzz/src/runtime/stack.c:299:13
    #4 0x57732b in ts_stack_pop_count /src/octofuzz/src/runtime/stack.c:427
    #5 0x55f96c in parser__reduce /src/octofuzz/src/runtime/parser.c:652:24
    #6 0x55bf87 in parser__advance /src/octofuzz/src/runtime/parser.c:1123:38
    #7 0x55adca in parser_parse /src/octofuzz/src/runtime/parser.c:1234:9
    #8 0x55352f in ts_document_parse_with_options /src/octofuzz/src/runtime/document.c:137:16
    #9 0x5182eb in LLVMFuzzerTestOneInput /src/octofuzz/test/fuzz/fuzzer.cc:21:3
    #10 0x5a470e in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/libfuzzer/FuzzerLoop.cpp:463:13
    #11 0x584145 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) /src/libfuzzer/FuzzerDriver.cpp:273:6
    #12 0x58f3df in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/libfuzzer/FuzzerDriver.cpp:689:9
    #13 0x5837e8 in main /src/libfuzzer/FuzzerMain.cpp:20:10
    #14 0x7f184626682f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

Indirect leak of 128 byte(s) in 1 object(s) allocated from:
    #0 0x4dd168 in malloc /src/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:88
    #1 0x568d9b in ts_malloc /src/octofuzz/src/runtime/alloc.h:49:18
    #2 0x56aa4b in ts_tree_pool_allocate /src/octofuzz/src/runtime/tree.c:151:12
    #3 0x56abf8 in ts_tree_make_leaf /src/octofuzz/src/runtime/tree.c:167:18
    #4 0x56dd04 in ts_tree_make_node /src/octofuzz/src/runtime/tree.c:378:18
    #5 0x55fbd0 in parser__reduce /src/octofuzz/src/runtime/parser.c:665:20
    #6 0x55bf87 in parser__advance /src/octofuzz/src/runtime/parser.c:1123:38
    #7 0x55adca in parser_parse /src/octofuzz/src/runtime/parser.c:1234:9
    #8 0x55352f in ts_document_parse_with_options /src/octofuzz/src/runtime/document.c:137:16
    #9 0x5182eb in LLVMFuzzerTestOneInput /src/octofuzz/test/fuzz/fuzzer.cc:21:3
    #10 0x5a470e in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/libfuzzer/FuzzerLoop.cpp:463:13
    #11 0x584145 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) /src/libfuzzer/FuzzerDriver.cpp:273:6
    #12 0x58f3df in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/libfuzzer/FuzzerDriver.cpp:689:9
    #13 0x5837e8 in main /src/libfuzzer/FuzzerMain.cpp:20:10
    #14 0x7f184626682f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

Indirect leak of 128 byte(s) in 1 object(s) allocated from:
    #0 0x4dd168 in malloc /src/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:88
    #1 0x568d9b in ts_malloc /src/octofuzz/src/runtime/alloc.h:49:18
    #2 0x56aa4b in ts_tree_pool_allocate /src/octofuzz/src/runtime/tree.c:151:12
    #3 0x56abf8 in ts_tree_make_leaf /src/octofuzz/src/runtime/tree.c:167:18
    #4 0x56dd04 in ts_tree_make_node /src/octofuzz/src/runtime/tree.c:378:18
    #5 0x56e098 in ts_tree_make_error_node /src/octofuzz/src/runtime/tree.c:396:18
    #6 0x567112 in parser__recover_to_state /src/octofuzz/src/runtime/parser.c:998:21
    #7 0x561133 in parser__recover /src/octofuzz/src/runtime/parser.c:1039:11
    #8 0x55c6c6 in parser__advance /src/octofuzz/src/runtime/parser.c:1144:11
    #9 0x55adca in parser_parse /src/octofuzz/src/runtime/parser.c:1234:9
    #10 0x55352f in ts_document_parse_with_options /src/octofuzz/src/runtime/document.c:137:16
    #11 0x5182eb in LLVMFuzzerTestOneInput /src/octofuzz/test/fuzz/fuzzer.cc:21:3
    #12 0x5a470e in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/libfuzzer/FuzzerLoop.cpp:463:13
    #13 0x584145 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) /src/libfuzzer/FuzzerDriver.cpp:273:6
    #14 0x58f3df in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/libfuzzer/FuzzerDriver.cpp:689:9
    #15 0x5837e8 in main /src/libfuzzer/FuzzerMain.cpp:20:10
    #16 0x7f184626682f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

Indirect leak of 128 byte(s) in 2 object(s) allocated from:
    #0 0x4dd380 in calloc /src/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:97
    #1 0x5737d4 in ts_calloc /src/octofuzz/src/runtime/alloc.h:58:18
    #2 0x573934 in array__grow /src/octofuzz/src/runtime/array.h:93:24
    #3 0x57732b in stack__iter /src/octofuzz/src/runtime/stack.c:299:13
    #4 0x57732b in ts_stack_pop_count /src/octofuzz/src/runtime/stack.c:427
    #5 0x55f96c in parser__reduce /src/octofuzz/src/runtime/parser.c:652:24
    #6 0x567fa1 in parser__do_all_potential_reductions /src/octofuzz/src/runtime/parser.c:837:7
    #7 0x562808 in parser__handle_error /src/octofuzz/src/runtime/parser.c:884:3
    #8 0x55c2c0 in parser__advance /src/octofuzz/src/runtime/parser.c:1162:7
    #9 0x55adca in parser_parse /src/octofuzz/src/runtime/parser.c:1234:9
    #10 0x55352f in ts_document_parse_with_options /src/octofuzz/src/runtime/document.c:137:16
    #11 0x5182eb in LLVMFuzzerTestOneInput /src/octofuzz/test/fuzz/fuzzer.cc:21:3
    #12 0x5a470e in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/libfuzzer/FuzzerLoop.cpp:463:13
    #13 0x584145 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) /src/libfuzzer/FuzzerDriver.cpp:273:6
    #14 0x58f3df in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/libfuzzer/FuzzerDriver.cpp:689:9
    #15 0x5837e8 in main /src/libfuzzer/FuzzerMain.cpp:20:10
    #16 0x7f184626682f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

SUMMARY: AddressSanitizer: 1856 byte(s) leaked in 18 allocation(s).
```

</details>

Testcases which trigger the leak for bash (`hexdump`ed):
```
00000000  3b 2a 3b 74 3b 2a 3b 74  3b 79 3b 2a 2a 3b 64     |;*;t;*;t;y;**;d|
```

and

```
00000000  26 3c 3c 3c 2d 3b 8f 26  3b 01 26 1a 0a 0a 3b     |&<<<-;.&;.&...;|
```

Bisecting the recent history, the leak seems to be introduced in 134c455b as part of #128

**Comment by maxbrunsfeld** [comment]
@philipturnbull I believe this was probably a legitimate bug; someone reported an Atom crash that I was able to trace down to a bug introduced in that commit https://github.com/tree-sitter/tree-sitter/commit/134c455b80782a9bd2ab76eb6e03e33dfb0768e8. I believe I have fixed it via https://github.com/tree-sitter/tree-sitter/commit/16a45d4aa439e543e7e355c57a7fb15767490efe, but I have not yet added automated tests covering this problem. I'd still be interested in pairing and learning to use the leak sanitizer.
